lib/Dialect/FIRRTL/Transforms/ExpandWhens.cpp

```diff
@@ -0,0 +1,392 @@
+      return value;
+    // 'and' the value with the current condition.
+    return OpBuilder(op).createOrFold<AndPrimOp>(
+        condition.getLoc(), condition.getType(), condition, value);
+  }
+
+  /// Take two connection operations and merge them in to a new connect under a
+  /// condition.  Destination of both connects should be `dest`.
+  Operation *flattenConditionalConnections(OpBuilder &b, Location loc,
+                                           Value dest, Value cond,
+                                           Operation *whenTrueConn,
+                                           Operation *whenFalseConn) {
+    auto whenTrue = getConnectedValue(whenTrueConn);
+    auto whenFalse = getConnectedValue(whenFalseConn);
+    auto newValue = b.createOrFold<MuxPrimOp>(loc, whenTrue.getType(), cond,
+                                              whenTrue, whenFalse);
+    auto newConnect = b.create<ConnectOp>(loc, dest, newValue);
+    whenTrueConn->erase();
+    whenFalseConn->erase();
+    return newConnect;
+  }
+
+private:
+  /// Map of destinations and the operation which is driving a value to it in
+  /// the current scope. This is used for resolving last connect semantics, and
+  /// for retrieving the responsible connect operation.
+  llvm::MapVector<Value, Operation *> &scope;
+
+  /// The current wrapping condition. If null, we are in the outer scope.
+  Value condition;
+};
+} // namespace
+
+LogicalResult ExpandWhensVisitor::run(FModuleOp module) {
+  llvm::MapVector<Value, Operation *> outerScope;
+  ExpandWhensVisitor(outerScope, nullptr).visitDecl(module);
+
+  // Check for any incomplete initialization.
+  LogicalResult result = success();
+  for (auto destAndConnect : outerScope) {
+    auto connect = std::get<1>(destAndConnect);
+    if (connect)
+      continue;
+    // This pass has failed.
+    result = failure();
+    auto dest = std::get<0>(destAndConnect);
+    auto op = dest.getDefiningOp();
+
+    // Module ports.
+    if (auto arg = dest.dyn_cast<BlockArgument>()) {
+      // Get the module ports and get the name.
+      SmallVector<ModulePortInfo> ports = module.getPorts();
+      module.emitError("module port ")
+          << ports[arg.getArgNumber()].name << " not fully initialized.";
+      continue;
+    }
+
+    // Instance operation.
+    if (auto instance = dyn_cast<InstanceOp>(op)) {
+      auto result = dest.cast<OpResult>();
+      instance->emitError("instance port ")
+          << instance.getPortName(result.getResultNumber())
+          << " not fully initialized.";
+      continue;
+    }
+
+    // Memory operation.
+    // TODO: Memories currently are capable of returning bundles. The pass is
+    // not capable of tracking individual fields in a bundle, and can only
+    // detect connections to the bundle as a whole.  After issue #888 is solved,
+    // LowerTypes will transform memories to only return ground types.
+    if (auto memory = dyn_cast<MemOp>(op)) {
+      continue;
+    }
+
```

Comment:
None of the new initialization checking will work properly when there are sinks with bundle types.  I think all bundles, except for the ones from memory ports, are removed by the LowerTypes pass.